src/Http/Http.Abstractions/src/Extensions/UseMiddlewareExtensions.cs

```diff
@@ -209,19 +210,70 @@ public RequestDelegate CreateMiddleware(RequestDelegate next)
             }
         }
 
+        // Performance optimization: Precompute and cache the key results for each parameter
+        var precomputedKeys = new object?[parameters.Length];
+        for (var i = 1; i < parameters.Length; i++)
+        {
+            var hasServiceKey = TryGetServiceKey(parameters[i], out object? key);
+            precomputedKeys[i] = key;
+        }
+
         return (middleware, context, serviceProvider) =>
         {
             var methodArguments = new object[parameters.Length];
             methodArguments[0] = context;
             for (var i = 1; i < parameters.Length; i++)
             {
-                methodArguments[i] = GetService(serviceProvider, parameters[i].ParameterType, methodInfo.DeclaringType!);
+                var key = precomputedKeys[i];
+                var parameterType = parameters[i].ParameterType;
+                var declaringType = methodInfo.DeclaringType!;
+
+                methodArguments[i] = key == null ? GetService(serviceProvider, parameterType, declaringType) : GetKeyedService(serviceProvider, key, parameterType, declaringType);
             }
 
             return (Task)methodInfo.Invoke(middleware, BindingFlags.DoNotWrapExceptions, binder: null, methodArguments, culture: null)!;
         };
     }
 
+    private static bool TryGetServiceKey(ParameterInfo parameterInfo, out object? key)
```

Comment:
```suggestion
    private static bool TryGetServiceKey(ParameterInfo parameterInfo, [MemberNotNullWhen(true)] out object? key)
```

Reply:
@captainsafia, thanks for the feedback. I've implemented it together with some other changes. Would've used the "Commit suggestion" otherwise.